Sys bindings update

matthiascy opened the pull request:
Hi @Twinklebear,

I updated the sys bindings from v3.12.1 to v3.13.5, device/scene creation now returns a `Result<Device/Scene, Error>` to make sure that the device/scene is created successfully. Sys binding generation script is also updated due to rust-lang/rust-bindgen#1812.

By the way, embree4 is released yeasterday. Do we plan to update? Probably we should finish what left to implement for embree3. I'm quite experienced with embree3, maybe I can help?

matthiascy committed:
update sys binding generation script

matthiascy committed:
update sys bindings generated against v3.13.5

matthiascy committed:
thin RTCError wrapper & allow device creation to fail

matthiascy committed:
fix doc tests

matthiascy committed:
format sys.rs

matthiascy committed:
remove RTCError wrapper

matthiascy committed:
fix typos

matthiascy committed:
resolve Twinklebear/embree-rs#23

matthiascy committed:
remove Arc from Scene and Device creation

matthiascy commented:
Commit bf678b0 tries to resolve #3

matthiascy committed:
update doc test

matthiascy committed:
refactor Scene creation as device method

To underline that a scene is bound to a specific device,
the scene creation function is refactored as a member function
of device.

matthiascy committed:
update doc test

matthiascy committed:
start refactoring buffer

matthiascy committed:
tmp commit

matthiascy committed:
tmp commit

matthiascy committed:
refactor Buffer/Geometry

- Buffer can be referenced by creating BufferView or BufferViewMut
- Different part of Buffer can be referenced by creating BufferSlice
- BufferView and BufferViewMut can be created from BufferSlice
- Shared data between different geometry objects now resides in BufferGeometry

matthiascy committed:
add rustfmt.toml configuration file

Twinklebear commented:
Thanks @matthiascy ! I'll take a look through the changes this week. 

As for Embree4, we should update to it, but I think it'll be pretty easy for CPU support since they've added the GPU support without making many API breaking changes. So CPU code can be updated with just a few smaller changes.

Adding support for GPU ray tracing w/ Embree 4 would be super cool, but pretty challenging. I'm not sure how we'd be able to build Rust code using Embree on the GPU to SYCL/SPV kernels that we can run with Intel's Level Zero runtime. The Intel GPU RT intrinsics used by Embree are some extensions in SYCL/OpenCL, so while there's stuff like rust-gpu (https://github.com/EmbarkStudios/rust-gpu) , I don't know how much we could re-purpose or use it since their target is Vulkan shaders. Certainly something to think about though.

matthiascy committed:
minimal example & add all examples in workspace Twinklebear/embree-r#21

matthiascy committed:
update CI embree version

matthiascy committed:
try to include set shared buffer

matthiascy committed:
refactor buffer & slice

matthiascy committed:
update minimal example

matthiascy committed:
fix some clippy warnings

Twinklebear commented on src/geometry.rs:
```diff
@@ -1,20 +1,405 @@
+                }
+            }
+            BufferSlice::Internal { .. } => {
+                eprint!("Internally managed buffer cannot be shared!");
```
I think we should re-word this slightly because "internally managed buffer" could also be read as buffers made via `Buffer::new` which are also memory "internally managed" by Embree.

Maybe `BufferSlice::Internal` could be `BufferSlice::GeometryLocal` ? Then the error here would be about trying to share a geometry-local buffer.

matthiascy replied:
Renamed as `GeometryLocal`

Twinklebear commented on src/instance.rs:
```diff
@@ -18,14 +14,12 @@ impl Instance {
-            handle: h,
-            scene: scene,
-        })
+        Arc::new(Instance { handle: h, scene })
```
Do you also plan to or should we remove the Arc from Instance as well? I think it can get the same treatment to use rtcRetain/rtcRelease as the other objects (I may just be reviewing this before you've gotten to it, if so no problem :) )

matthiascy replied:
I removed the Arc from instance, scene now managed internall by embree.

Twinklebear commented on src/scene.rs:
```diff
@@ -1,33 +1,58 @@
     pub(crate) handle: RTCScene,
-    pub(crate) device: Arc<Device>,
+    pub(crate) device: Device,
     geometry: HashMap<u32, Arc<dyn Geometry>>,
```
Maybe we can't remove the Arc around the geometry here because we're using dynamic dispatch/trait objects?

Twinklebear reviewed (changes requested):
Thanks a lot @matthiascy this looks awesome! I put some notes through the PR about smaller things, and just had some higher-level questions/notes as well:

- Do you also plan or think we can remove the `Arc<dyn Geometry>` from the scene? Maybe to be just an `Rc`? Or remove entirely? It may not be possible since we're using trait objects w/ runtime types/inheritance, but I'm a bit rusty
- With the improvements to the way buffers are set on objects, should we even keep the `::unanimated` implementations for the different Geometries? I don't see those being a common codepath for real applications which would likely be passing `BufferSlice::User` to us, so I think we could just remove them. Removing them also results in the API following Embree's C API more closely, where you make a geometry, then attach buffers to it, commit it, etc.
- The macOS failure looks like it just needs to have the `cp` command change to `cp -r` to copy the dirs properly.
- The Linux failure looks like it's just formatting

matthiascy commented:
> * Do you also plan or think we can remove the `Arc<dyn Geometry>` from the scene? Maybe to be just an `Rc`? Or remove entirely? It may not be possible since we're using trait objects w/ runtime types/inheritance, but I'm a bit rusty

Myabe we can remove `dyn Geometry` entirely. There is not need to keep `Geometry` trait, knowing that most methods are implemented on `BufferGeometry`. We can rename `BufferGeometry` as `Geometry` then the scene accepts references(`&` or `Rc` or `Arc`) to the `Geometry`. As concrete geometry types deref to `&Geometry`,  `attach_geometry` can accept `&Geometry` instead of using trait objects. 
For the moment, `BufferGeometry` is non-clonable and non-copyable.
- If we use shared reference `&`, then we are enforing lifetime dependency between `Scene` and geometry object, but according to Embree, after committing a geometry object, it can then be released, so this does not describe perfectly the relationship between geometries and a scene.
- If we use `Rc` and `Arc`: because that the `BufferGeometry` maintains information like attached buffers, we need to put these fields behind a `Arc<Mutex<..>>` to make it sharable across threads but only one can modify it at a time.

Or following the idea of encapsulating some fields inside `Arc<Mutex<...>>`, we can make `BufferGeometry` clonable, then when attaching to the scene, we still use `&Geometry` as input prameter but clone it to store inside the scene.
```rust
attach_geometry(&mut self, mesh: &Geometry) -> u32 {
   ...
   self.geometry.insert(id, mesh.clone());
}
```

The `Scene` can also be implemented this way, clonable but control the access to some field, then we allow sharing `Scene` between threads(this is needed, it is thread-safe to call `intersect/ocluded` function from multiple threads). Don't know if it would be possible to discard the field of attached geometries. If so, `Scene` would be even more simple.
```rust
pub struct Scene {
    pub(crate) handle: RTCScene,
    pub(crate) device: Device,
    geometry: Arc<Mutex<HashMap<u32, BufferGeometry>>>,
}
```

> * With the improvements to the way buffers are set on objects, should we even keep the `::unanimated` implementations for the different Geometries? I don't see those being a common codepath for real applications which would likely be passing `BufferSlice::User` to us, so I think we could just remove them. Removing them also results in the API following Embree's C API more closely, where you make a geometry, then attach buffers to it, commit it, etc.

I think we can just remove it.

Last thing, should we keep track of the mapped range of each Buffer to make sure there is no overlap?
```rust
#[derive(Debug)]
pub struct Buffer {
    pub(crate) device: Device,
    pub(crate) handle: RTCBuffer,
    pub(crate) size: BufferSize,
    pub(crate) mapped_ranges: Mutex<Vec<Range>>
}
```

matthiascy committed:
remove Geometry trait

matthiascy commented on src/buffer.rs:
```diff
@@ -1,151 +1,410 @@
+            BufferSlice::User {
+                ptr, offset, size, ..
+            } => {
+                // TODO(yang): should we allow this?
```
Now returns an error of invalid operation: "Creating a view from a user owned/borrowed memory is not allowed!"

matthiascy committed:
try to port dynamic scene

matthiascy committed:
add some geometry related methods

matthiascy committed:
dynamic scene

matthiascy committed:
docs

matthiascy committed:
remove geometry state

matthiascy committed:
set geometry mask

matthiascy committed:
implementation of geometry related callback functions

matthiascy committed:
update build test scripts & fix examples

matthiascy committed:
fix minimal

matthiascy committed:
put user/instance related methods under its own struct

matthiascy committed:
started adopting instance example

matthiascy commented:
Hi @Twinklebear, I adopted the dynamic scene from official repo:
![Screenshot from 2023-02-20 22-58-22](https://user-images.githubusercontent.com/8416512/220208178-8367d99c-c00a-4ce4-a611-810a841368aa.png)

Now most callbacks are implemented (not tested yet), still need some improvements. I'll try to adopt all the examples from official repo in the following weeks.

matthiascy committed:
user geometry payloads for callbacks

matthiascy committed:
instancing example done

matthiascy committed:
add quaternion decomposition

matthiascy committed:
some intersection methods

matthiascy committed:
separate subdivision & impl safe ray packet access in callback functions

matthiascy committed:
implement all callback functions

matthiascy committed:
remove portable simd

matthiascy committed:
remove user geometry

matthiascy committed:
clean displacement callback

matthiascy committed:
displacement geometry example

matthiascy committed:
temp update

matthiascy committed:
reduce code duplication

matthiascy committed:
remove geomtry directory & update workflows to use nightly toolchain

matthiascy committed:
update code formatting task

matthiascy committed:
update examples

matthiascy committed:
tiled image

matthiascy committed:
update tiled image for displacement geometry

matthiascy committed:
trait for extending ray

Twinklebear commented:
Awesome, thanks for all your work @matthiascy ! I should have time to review the PR in the next few days.

matthiascy committed:
poor man's inheritance for ray

matthiascy committed:
remove AsRay/AsRayHit

matthiascy committed:
fix instancing & intersect filter

matthiascy committed:
update some callback functions

matthiascy committed:
api ergonomics improvement & performace boost & intersection filter

matthiascy committed:
refactor ray/hit stream

matthiascy committed:
add miri test; fix memory leak & incorrect memory deallocation

matthiascy committed:
update ci

matthiascy committed:
remove marker

matthiascy committed:
remove marker from aligned vector & add aligned array

matthiascy committed:
egui integration

matthiascy committed:
examples performance boost & different shading method

matthiascy commented:
Updates:

1. Replaced old image to `TiledImage` to enable parallel rendering per image tile in examples.
2. Intergrated egui to enable UI displaying.
3. Fixed window can't be resized.
4. Implemented `EyeLight`, `UV`, `Normal`, `CPUCycles`, `GeometryID` and `GeometryPrimitiveID` shading mode. (missing `Occlusion`, `TexCoords`, `TexCoordsGrid`, `AmbientOcclusion`)

![Screenshot from 2023-03-13 11-11-23](https://user-images.githubusercontent.com/8416512/224674112-a99f517d-7884-4267-ba16-648421c12d82.png)
![Screenshot from 2023-03-13 11-11-30](https://user-images.githubusercontent.com/8416512/224674119-017e8c33-1dfa-4545-a96c-ad5467bac365.png)
![Screenshot from 2023-03-13 11-11-38](https://user-images.githubusercontent.com/8416512/224674123-921952f9-b13c-4645-b754-5ce0925a0f10.png)
![Screenshot from 2023-03-13 11-11-49](https://user-images.githubusercontent.com/8416512/224674126-3a4c38e0-4665-4d6f-8d1b-eb5696de9437.png)
![Screenshot from 2023-03-13 11-11-56](https://user-images.githubusercontent.com/8416512/224674129-d824a515-4671-4590-bb6d-4942b1e73c10.png)
![Screenshot from 2023-03-13 11-12-02](https://user-images.githubusercontent.com/8416512/224674130-e1e37fd9-5e03-4af0-a45b-66938e65ae2d.png)
![Screenshot from 2023-03-13 11-12-06](https://user-images.githubusercontent.com/8416512/224674133-51c9acc2-15bd-4bf9-abfd-0a83657024cc.png)

Twinklebear commented:
Thanks again @matthiascy ! Sorry that my "next few days" note above has turned out to be untrue, I've been tied up with some other projects.

Do you plan to add those other renderers or additional features for this PR? I was also waiting a bit as I saw you were still pushing new stuff.

I have a paper deadline March 31 so I may not have enough time to look through everything properly until after that, if there were additional things you wanted to add. We could look at updating to Embree 4 as well, the changes on the CPU side are not too big (though there are some breaking changes).

Twinklebear commented:
Hi @matthiascy , I just wanted to check in if this was ready for a review through yet? It sounded like you planned to add a lot more to get up to 3.13.5. No pressure if you're busy with other stuff, I just wanted to make sure you weren't blocked waiting for me to review.

matthiascy commented:
Hi @Twinklebear, your review is not causing any delays or blocking my work. However, because of other projects, I now have limited time to work on it. I'll try to finish it once my present project is done. :)

Twinklebear commented:
No problem @matthiascy ! No pressure, and thanks for all your hard work so far 😄
